### PR TITLE
Proposal: move all python dependencies to cfgov-refresh

### PR DESCRIPTION
**Pull request by rosskarchner**
Per "[New Build Approach](https://cfpb.github.io/cfgov-refresh/whats-new/)", we want this repo to reflect all of our Python dependencies. This makes all of the previous "optional" dependencies that were stored in a separate (private) repo are now expressed in requirements/optional.txt.

If you look at optional.txt, you'll see this note:
```
# some of these dependencies live on a private git server. If you have the address,
# configure git with:
# git config --global url.https://<private hostname>/.insteadOf https://private.repo/
```
That's how it works: with that git config command, we're telling git to always replace private.repo with our internal git server, and that also applies when git is invoked by pip. (internal) developers who want to install the optional apps, and our own deployment pipeline, can be configured to make that transformation as part of the process.

We should make sure that tests always run, and the site is basically functional, WITHOUT these dependencies installed. This is why we have the "[OPTIONAL_APPS](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/cfgov/settings/base.py#L66)" system.

## Review

Particualrly interested in thoughts from:
- @grapesmoker 
- @willbarton 

## Checklist

* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

### CODE REVIEW

**Comment by Scotchester** [comment]
Neat idea!

May I suggest grouping the `private.repo` dependencies together and the `github.com` dependencies together in the optional requirements file?

File: requirements/optional.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# some of these dependencies live on a private git server. If you have the address,
+# configure git with:
+# git config --global url.https://<private hostname>/.insteadOf https://private.repo/
+
+git+https://private.repo/CFPB/cfgov-selfregistration.git@v1.1#egg=selfregistration
```
Comment (vccabral) [inline requirements/optional.txt]: How do you imagine people are going to install these things? Is that documented anywhere?

**Comment by rosskarchner** [comment]
@Scotchester That's a good idea

File: requirements/optional.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# some of these dependencies live on a private git server. If you have the address,
+# configure git with:
+# git config --global url.https://<private hostname>/.insteadOf https://private.repo/
```
Comment (willbarton) [inline requirements/optional.txt]: This is clever... but I'm still not convinced it's a better solution than just defining our private requirements in the same places as the rest of our private setup details.

Comment (rosskarchner) [inline requirements/optional.txt]: @willbarton I think you proposed making cfgov-refresh into an installable python package, and maintaining a requirements.txt in a private repo that references a given cfgov-refresh version, along with these private repos, releasing versions of *that* repo. Let's call this new repo cfgov-everything

requirements.txt in cfgov-everything (master) might look like:

```
git+https://github.com/cfpb/cfgov-refresh.git@3.5.1
git+https://secret/CFPB/cfgov-selfregistration.git@v1.1#egg=selfregistration
...
```
Or maybe like this
```
git+https://github.com/cfpb/cfgov-refresh.git@master
git+https://secret/CFPB/cfgov-selfregistration.git@v1.1#egg=selfregistration
...
```

In the first version, On our build server, we deploy the master version of cfgov-everything, except for cfgov-refresh: we ignore the pinned version and deploy master of *that*. In the second, master of cfgov-everything implies master of cfgov-refresh, and when we create a release, we have to create releases of both repos, and update requirements.txt in the release branch of cfgov-everything to point to the new release of cfgov-refresh.

It's not terrible, especially if we automate things.

I think this is simpler, though, and I see advantages in having a single repo that describes the (code) state of the website. For one thing, it'll make it easier to communicate with groups like release management and security.

But honestly, it's just the difference between "this is cf.gov" vs. "this and this are cf.gov" both are superior to the current setup, of course ;)

Comment (willbarton) [inline requirements/optional.txt]: @rosskarchner Ok. Thinking long-term, doing it the way you're proposing here might encourage more things to be public and open source than private anyway.

I'm sold on this for now. Let's have one source of truth, and let this be it.

:+1:

**Comment by willbarton** [comment]
@rosskarchner I think we can merge this once you group the repos together like @Scotchester suggested.

**Comment by rosskarchner** [comment]
I'm thinking about the timing-- maybe we let dynamic disclosures release the "old" way, and then make the switch.

**Comment by willbarton** [comment]
@rosskarchner that makes sense to me.

**Comment by rosskarchner** [comment]
Closing this for now

**Comment by rosskarchner** [comment]
Bringing this back to life: I'll address @Scotchester's feedback and get this in shape for the next release.

File: requirements/optional-private.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# These dependencies live on a private git server. If you have the address,
+# configure git with:
+# git config --global url.https://<private hostname>/.insteadOf https://private.repo/
+
+git+https://private.repo/CFPB/cfgov-selfregistration.git@v1.1#egg=selfregistration
+git+https://private.repo/CFPB/Picard.git@v1.3#egg=picard
+git+https://private.repo/CFPB/django-college-cost-comparison.git@v1.2.2#egg=comparisontool
+git+https://private.repo/CFPB/knowledgebase.git@v2.1.0#egg=knowledgebase
+git+https://private.repo/CFPB/agreement_database.git@v2.2.1#egg=agreements
+git+https://private.repo/CCDB4/CCDB-content.git@v1.2.2#egg=complaintdatabase
```
Comment (higs4281) [inline requirements/optional-private.txt]: this ccdb import should be deleted

File: requirements/optional-public.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+git+https://github.com/cfpb/django-hud.git@1.2#egg=django-hud
+git+https://github.com/cfpb/owning-a-home-api.git@v0.9.7#egg=owning-a-home-api
+git+https://github.com/cfpb/retirement.git@0.5.0#egg=retirement
+git+https://github.com/cfpb/regulations-site.git@2.1.3#egg=regulations
+git+https://github.com/cfpb/regulations-core.git@1.2.3#egg=regcore
+git+https://github.com/cfpb/college-costs.git@2.2.2#egg=college-costs
+
```
Comment (higs4281) [inline requirements/optional-public.txt]: should add the public ccdb repo:

`git+https://github.com/cfpb/complaint.git@v1.2.3#egg=complaintdatabase`

**Comment by higs4281** [comment]
@rosskarchner I have a version ready that uses the public ccdb repo, resolves master conflicts and passes tests, if you want it.

**Comment by rosskarchner** [comment]
@higs4281 Push it!

**Comment by higs4281** [comment]
Can we take this opportunity for a security update?  
django 1.8.15 catches a CSRF vulnerability: https://www.djangoproject.com/weblog/2016/sep/26/security-releases/#s-cve-2016-7401-csrf-protection-bypass-on-a-site-with-google-analytics

**Comment by rosskarchner** [comment]
Yeah, go ahead and add it.

**Pull request merged**